### PR TITLE
Bench driven modifs

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/boost/simd/arch/common/simd/function/is_lessgreater.hpp [hunk 1]
```diff
@@ -59,9 +59,9 @@ namespace boost { namespace simd { namespace ext
                           )
   {
     BOOST_FORCEINLINE auto operator() ( A0 a0, A0 a1) const BOOST_NOEXCEPT
-   -> decltype(is_not_equal(a0, a1))
+      -> decltype(logical_and(is_ord(a0,a1), is_not_equal(a0, a1)))
```
Comment (jfalcou) [inline include/boost/simd/arch/common/simd/function/is_lessgreater.hpp]: Commenting here for the whole file :
- the result_type is obviously as_logical_t<A0>, no need for auto-decltype. This is probably a remnant of the old implementation
- Pass a0 and a1 by const-ref or MSVC will cry.

File: bench/function/scalar/abs_s.cpp [hunk 2]
```diff
@@ -16,7 +16,7 @@ DEFINE_SCALAR_BENCH(scalar_abs_s, bs::saturated_(bs::abs));
 
 int main(int argc, char** argv) {
   nsb::parse_args(argc, argv);
-  nsb::for_each<scalar_abs_s, NS_BENCH_NUMERIC_TYPES>(-10, 10);
+  nsb::for_each<scalar_abs_s, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/scalar/abs_s.cpp]: why the change ?

File: bench/function/scalar/genmask.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#include <boost/simd/function/simd/genmask.hpp>
+
+namespace nsb = ns::bench;
+namespace bs =  boost::simd;
+
+DEFINE_SCALAR_BENCH(scalar_genmask, bs::genmask);
+
+DEFINE_BENCH_MAIN()
+{
+  nsb::for_each<scalar_genmask, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/scalar/genmask.cpp]: genmak can be benchmarked for all types

File: bench/function/scalar/shift_left.cpp [hunk 4]
```diff
@@ -8,13 +8,26 @@
...
+  }
+};
+
+DEFINE_SCALAR_BENCH(scalar_shift_left1, shlN<1>());
+DEFINE_SCALAR_BENCH(scalar_shift_left2, shlN<2>());
 
 DEFINE_BENCH_MAIN()
 {
-  nsb::for_each<scalar_shift_left, NS_BENCH_IEEE_TYPES>(-10, 10);
+  nsb::for_each<scalar_shift_left1, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/scalar/shift_left.cpp]: this can be benchmarked with all types

File: bench/function/simd/abs_s.cpp [hunk 5]
```diff
@@ -17,7 +17,7 @@ DEFINE_SIMD_BENCH(simd_abs_s, bs::saturated_(bs::abs));
 
 int main(int argc, char** argv) {
   nsb::parse_args(argc, argv);
-  nsb::for_each<simd_abs_s, NS_BENCH_NUMERIC_TYPES>(-10, 10);
+  nsb::for_each<simd_abs_s, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/simd/abs_s.cpp]: again, this can use all types

File: bench/function/simd/genmask.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+#include <boost/simd/pack.hpp>
+
+namespace nsb = ns::bench;
+namespace bs =  boost::simd;
+
+DEFINE_SIMD_BENCH(simd_genmask, bs::genmask);
+
+DEFINE_BENCH_MAIN()
+{
+  nsb::for_each<simd_genmask, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/simd/genmask.cpp]: again, all types

File: bench/function/simd/shift_left.cpp [hunk 7]
```diff
@@ -8,14 +8,26 @@
...
+  }
+};
+
+DEFINE_SIMD_BENCH(scalar_shift_left1, shlN<1>());
+DEFINE_SIMD_BENCH(scalar_shift_left2, shlN<2>());
 
 DEFINE_BENCH_MAIN()
 {
-  nsb::for_each<simd_shift_left, NS_BENCH_IEEE_TYPES>(-10, 10);
+  nsb::for_each<scalar_shift_left1, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/simd/shift_left.cpp]: all types there too

File: bench/function/simd/shift_right.cpp [hunk 8]
```diff
@@ -8,14 +8,26 @@
...
+  }
+};
+
+DEFINE_SIMD_BENCH(scalar_shift_right1, shlN<1>());
+DEFINE_SIMD_BENCH(scalar_shift_right2, shlN<2>());
 
 DEFINE_BENCH_MAIN()
 {
-  nsb::for_each<simd_shift_right, NS_BENCH_IEEE_TYPES>(-10, 10);
+  nsb::for_each<scalar_shift_right1, NS_BENCH_IEEE_TYPES>(-10, 10);
```
Comment (jfalcou) [inline bench/function/simd/shift_right.cpp]: all types there too

File: include/boost/simd/arch/common/generic/function/meanof.hpp [hunk 9]
```diff
@@ -38,13 +37,12 @@ namespace boost { namespace simd { namespace ext
...
                           )
   {
-    using result_type = A0;
     BOOST_FORCEINLINE A0 operator() ( A0 const& a0, A0 const& a1) const BOOST_NOEXCEPT
     {
       A0 m = min(a0, a1);
-      return if_else( bitwise_and(is_finite(a0), is_finite(a1)),//is_finite(a0) && is_finite(a1),
-                      m + (max(a0, a1)-m)*Half<result_type>(),
-                      average(a0, a1)
+      return if_else( logical_and(is_finite(a0), is_finite(a1)),//is_finite(a0) && is_finite(a1),
```
Comment (jfalcou) [inline include/boost/simd/arch/common/generic/function/meanof.hpp]: considering there is a large branch, this is maybe requried ot be split in scalar//simd for performance's sake.

File: include/boost/simd/arch/common/scalar/function/logical_xor.hpp [hunk 10]
```diff
@@ -28,7 +28,7 @@ namespace boost { namespace simd { namespace ext
   {
     BOOST_FORCEINLINE logical<A0> operator() ( A0 a0, A1 a1) const BOOST_NOEXCEPT
     {
-      return {(a0 != a1)};
+      return {(!a0 != !a1)};
```
Comment (jfalcou) [inline include/boost/simd/arch/common/scalar/function/logical_xor.hpp:31]: can you explain this change ?

**Comment by jfalcou** [comment]
Will merge as soon as the two point on logical_xor and meanof are answered

File: include/boost/simd/function/simd/ldexp.hpp [hunk 11]
```diff
@@ -14,10 +14,4 @@
 #include <boost/simd/arch/common/generic/function/autodispatcher.hpp>
 #include <boost/simd/arch/common/simd/function/ldexp.hpp>
 
-#if defined(BOOST_HW_SIMD_X86_OR_AMD_AVAILABLE)
```
Comment (jfalcou) [inline include/boost/simd/function/simd/ldexp.hpp:17]: why is it removed ?

Comment (jtlap) [inline include/boost/simd/function/simd/ldexp.hpp:17]: because the new addings to the common/simd version make it obsolete

File: include/boost/simd/arch/common/scalar/function/logical_xor.hpp (hunk 10 above)
Comment (jtlap) [inline include/boost/simd/arch/common/scalar/function/logical_xor.hpp:31]: because it's an error logical_xor(1,2) must be false not true : they are both true

File: include/boost/simd/function/simd/ldexp.hpp (hunk 11 above)
Comment (jtlap) [inline include/boost/simd/function/simd/ldexp.hpp:17]: Do you man ldexp ? what about meanof ?
